src/geometry.cpp

```diff
@@ -536,6 +536,196 @@ extern "C" double openmc_calculate_optical_thickness(
   return optical_thickness;
 }
 
+extern "C" void openmc_get_voxel_cross_sections(Position voxel_min,
+  Position voxel_max, int num_samples, double* sigma_t_output,
+  double* sigma_a_output)
+{
+  double total_sigma_t = 0.0;
+  double total_sigma_a = 0.0;
+  SpatialBox voxel_box(voxel_min, voxel_max);
+  int64_t id = 1;
+  uint64_t seed = init_seed(id, STREAM_SOURCE);
+  int valid_samples = 0;
+
+#pragma omp parallel for reduction(                                            \
+    + : total_sigma_t, total_sigma_a, valid_samples)
+  for (int i = 0; i < num_samples; ++i) {
+    Position sampled_position = voxel_box.sample(&seed);
+
+    Particle p;
+    SourceSite site;
+    site.E = 1.0;
+    site.particle = ParticleType::neutron;
+    site.r = sampled_position;
+    site.u = {1.0, 0.0, 0.0}; // Direction doesn't matter for cross section
+    p.from_source(&site);
+    p.cell_born() = p.lowest_coord().cell;
+
+    exhaustive_find_cell(p);
+    p.event_calculate_xs();
+
+    total_sigma_t += p.macro_xs().total;
+    total_sigma_a += p.macro_xs().absorption;
+    valid_samples++;
+  }
+
+  if (valid_samples > 0) {
+    *sigma_t_output = total_sigma_t / valid_samples;
+    *sigma_a_output = total_sigma_a / valid_samples;
+  } else {
+    *sigma_t_output = 0.0;
+    *sigma_a_output = 0.0;
+  }
+}
+
+extern "C" void openmc_get_transport_operator_between_voxels(
+  Position start_voxel_min, Position start_voxel_max, Position end_voxel_min,
+  Position end_voxel_max, int num_rays, double* transport_operator_output)
+{
+  double total_transport = 0.0;
+  SpatialBox start_box(start_voxel_min, start_voxel_max);
+  SpatialBox end_box(end_voxel_min, end_voxel_max);
+  int64_t id = 1;
+  uint64_t seed = init_seed(id, STREAM_SOURCE);
+  int valid_samples = 0;
+
+#pragma omp parallel for reduction(+ : total_transport, valid_samples)
+
+  for (int i = 0; i < num_rays; ++i) {
+    Position start_sampled_position = start_box.sample(&seed);
+    Position end_sampled_position = end_box.sample(&seed);
+
+    double distance = (end_sampled_position - start_sampled_position).norm();
+    double optical_thickness = openmc_calculate_optical_thickness(
+      start_sampled_position, end_sampled_position);
+
+    double transport = std::exp(-optical_thickness) / (distance * distance);
```

Comment:
[performance] Avoid the sqrt and then squaring
```suggestion
    double ray = (end_sampled_position - start_sampled_position);
    double length_squared = ray.dot(ray);
    double optical_thickness = openmc_calculate_optical_thickness(
      start_sampled_position, end_sampled_position);

    double transport = std::exp(-optical_thickness) / length_squared;
```